### PR TITLE
Go Test: terraform-ecs-app examples

### PR DESCRIPTION
**Pull request by igorkotof**
- Web
- Worker

### CODE REVIEW

File: ecs-modules/ecs-service/outputs.tf [hunk 1]
```diff
@@ -12,3 +12,8 @@ output "cloudwatch_log_group" {
   description = "Cloudwatch Log group of ECS Service"
   value       = module.task.cloudwatch_log_group
 }
+
+output "cloudwatch_event_rule_id" {
+  description = "Cloudwatch event rule for ECS Scheduled Task"
+  value       = (var.app_type == "worker" && length(var.cloudwatch_schedule_expressions)>0) ? aws_cloudwatch_event_rule.this[0].id : "" 
+}
```
Comment (DimmKirr) [inline ecs-modules/ecs-service/outputs.tf]: New line please

File: test/src/examples_complete-web_test.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+// Test the Terraform module in examples/complete using Terratest.
+func TestExamplesCompleteWeb(t *testing.T) {
+	t.Parallel()
+	// randID := strings.ToLower(random.UniqueId())
+	// attributes := []string{randID}
+
+	rootFolder := "../../"
+	terraformFolderRelativeToRoot := "examples/complete-web"
+	varFiles := []string{"/Users/iharkotau/IT/PRJ/hazelops/github/hzo/terraform-aws-ecs-app/examples/complete-web/terraform.tfvars"}
```
Comment (DimmKirr) [inline test/src/examples_complete-web_test.go]: This doesn't look right. Needs to get the path dynamically

File: test/src/examples_complete-worker_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	terraform.InitAndApply(t, terraformOptions)
+
+	// Run `terraform output` to get the value of an output variable
+	vpcCidr := terraform.Output(t, terraformOptions, "vpc_cidr")
+	// Verify we're getting back the outputs we expect
+	assert.Equal(t, "10.30.0.0/16", vpcCidr)
+
+	// Run `terraform output` to get the value of an output variable
+	privateSubnetCidrs := terraform.OutputList(t, terraformOptions, "private_subnet_cidrs")
+	// Verify we're getting back the outputs we expect
```
Comment (DimmKirr) [inline test/src/examples_complete-worker_test.go:56]: New line

File: test/src/go.mod [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
+	k8s.io/api v0.20.6 // indirect
+	k8s.io/apimachinery v0.20.6 // indirect
+	k8s.io/client-go v0.20.6 // indirect
+	k8s.io/klog/v2 v2.4.0 // indirect
+	k8s.io/utils v0.0.0-20201110183641-67b214c5f920 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.0.3 // indirect
+	sigs.k8s.io/yaml v1.2.0 // indirect
+)
```
Comment (DimmKirr) [inline test/src/go.mod]: New line please

File: examples/complete-web/main.tf [hunk 5]
```diff
@@ -21,28 +120,28 @@ module "web_complete" {
...
   ]
 
   # Environment variables
   app_secrets = [
   ]
   environment = {
-    ENV      = var.env
-    APP_NAME = "App"
+    ENV       = var.env
+    APP_NAME  = "App"
```
Comment (DimmKirr) [inline examples/complete-web/main.tf]: Any reason to pass this? those parameters are passed anyway via the module

File: examples/complete-web/main.tf [hunk 6]
```diff
@@ -1,18 +1,117 @@
...
+    "10.30.12.0/23",
+    "10.30.14.0/23"
+  ]
+
+  private_subnets = [
+    "10.30.20.0/23",
+    "10.30.22.0/23",
+    "10.30.24.0/23"
+  ]
+  # database_subnets    = ["10.0.21.0/24", "10.0.22.0/24"]
```
Comment (DimmKirr) [inline examples/complete-web/main.tf]: Cleanup?

File: examples/complete-web/main.tf [hunk 7]
```diff
@@ -1,18 +1,117 @@
+module "vpc" {
+  source  = "registry.terraform.io/terraform-aws-modules/vpc/aws"
+  version = "~> 3.0"
+
+  name = "${var.env}-vpc"
+  cidr = "10.30.0.0/16"
```
Comment (DimmKirr) [inline examples/complete-web/main.tf]: Let's use more generic ip range please

File: .gitignore [hunk 8]
```diff
@@ -21,9 +21,23 @@ override.tf.json
 *_override.tf
 *_override.tf.json
 
+# Infra
+__pycache__
+terraform*.tfstate
+.terraform.lock.hcl
+*.tfvars
+*.wpvars
```
Comment (DimmKirr) [inline .gitignore]:
```suggestion
```

File: .gitignore [hunk 9]
```diff
@@ -21,9 +21,23 @@ override.tf.json
 *_override.tf
 *_override.tf.json
 
+# Infra
+__pycache__
```
Comment (DimmKirr) [inline .gitignore]:
```suggestion
```

File: .gitignore [hunk 10]
```diff
@@ -21,9 +21,23 @@ override.tf.json
...
+
 # Include override files you do wish to add to version control using negated pattern
 #
 # !example_override.tf
 
 # Include tfplan files to ignore the plan output of command: terraform plan -out=tfplan
 # example: *tfplan*
+
+**/.env
+.env*
```
Comment (DimmKirr) [inline .gitignore]:
```suggestion
```

File: examples/complete-web/main.tf [hunk 11]
```diff
@@ -1,18 +1,104 @@
...
+  ]
+
+  private_subnets = [
+    "10.30.20.0/23"
+  ]
+  manage_default_network_acl          = true
+  default_network_acl_name            = "${var.env}-${var.namespace}"
+}
+resource "aws_security_group" "default_permissive" {
+  name        = "${var.env}-default-permissive"
```
Comment (DimmKirr) [inline examples/complete-web/main.tf:22]: Let's add a more generic name

File: examples/complete-web/main.tf [hunk 12]
```diff
@@ -1,18 +1,104 @@
...
+    cidr_blocks = [
+      "0.0.0.0/0"
+    ]
+  }
+
+  tags = {
+    Terraform = "true"
+    Env       = var.env
+    Name      = "${var.env}-default-permissive"
+  }
```
Comment (DimmKirr) [inline examples/complete-web/main.tf]: Let's remove

Comment (DimmKirr) [inline examples/complete-web/main.tf]:
```suggestion
```

File: examples/complete-web/main.tf [hunk 13]
```diff
@@ -1,18 +1,104 @@
...
+}
+
+resource "aws_route53_record" "env_ns_record" {
+  zone_id = data.aws_route53_zone.root.id
+  name    = "${var.env}.${var.root_domain_name}"
+  type    = "NS"
+  //  ttl  = "172800"
+
+  // Fast TTL for dev
```
Comment (DimmKirr) [inline examples/complete-web/main.tf]:
```suggestion
```

File: examples/complete-worker/variables.tf [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  alb_security_groups  = [aws_security_group.default_permissive.id]
+  root_domain_name     = var.root_domain_name
+  zone_id              = aws_route53_zone.env_domain.id
+
+  image_id             = data.aws_ami.amazon_linux_ecs_generic.id
+  docker_registry      = var.docker_registry
+  docker_image_tag     = var.docker_image_tag
+
+  ecs_cluster_name     = module.ecs.cluster_name
+}
```
Comment (DimmKirr) [inline examples/complete-worker/variables.tf]: Let's simplify locals and not use it if it's just 1:1 value.

File: examples/complete-worker/versions.tf [hunk 15]
```diff
@@ -0,0 +1,9 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      #version = "~> 3.0,!= 3.14.0"
```
Comment (DimmKirr) [inline examples/complete-worker/versions.tf]: Remove version lock? or place back?
```suggestion
```

File: examples/proxy/data.tf [hunk 16]
```diff
@@ -0,0 +1,23 @@
+data "aws_availability_zones" "available" {}
```
Comment (DimmKirr) [inline examples/proxy/data.tf]: Let's name this example web-nginx-proxy?

File: examples/complete-worker/main.tf [hunk 17]
```diff
@@ -1,7 +1,79 @@
...
+    cidr_blocks = [
+      "0.0.0.0/0"
+    ]
+  }
+
+  tags = {
+    Terraform = "true"
+    Env       = var.env
+    Name      = "${var.env}-default-permissive"
+  }
```
Comment (DimmKirr) [inline examples/complete-worker/main.tf]:
```suggestion
```

File: examples/complete-worker/main.tf [hunk 18]
```diff
@@ -1,7 +1,79 @@
...
+  //  ttl  = "172800"
+
+  // Fast TTL for dev
+  ttl     = "60"
+  records = aws_route53_zone.env_domain.name_servers
+}
+
+resource "aws_route53_zone" "env_domain" {
+  name = "${var.env}.${var.root_domain_name}"
+}
```
Comment (DimmKirr) [inline examples/complete-worker/main.tf]: I don't think this is required for worker
```suggestion
```

File: examples/complete-worker/data.tf [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+  filter {
+    name   = "owner-alias"
+    values = ["amazon"]
+  }
+}
+
+data "aws_route53_zone" "root" {
+  name         = "${var.root_domain_name}."
+  private_zone = false
+}
```
Comment (DimmKirr) [inline examples/complete-worker/data.tf]: I don't think this is needed in a worker
```suggestion
```

File: examples/proxy/main.tf [hunk 20]
```diff
@@ -1,3 +1,92 @@
...
+  subject_alternative_names = [
+    "*.${local.env}.${local.root_domain_name}"
+  ]
+
+  zone_id = local.zone_id
+
+  tags = {
+    Name = "${var.env}.${var.root_domain_name}"
+  }
+}
```
Comment (DimmKirr) [inline examples/proxy/main.tf]: I would remove ACM from this example (not crucial, IMO)
```suggestion
```

File: ecs-modules/ecs-service/outputs.tf [hunk 21]
```diff
@@ -12,3 +12,8 @@ output "cloudwatch_log_group" {
   description = "Cloudwatch Log group of ECS Service"
   value       = module.task.cloudwatch_log_group
 }
+
+output "cloudwatch_event_rule_id" {
```
Comment (DimmKirr) [inline ecs-modules/ecs-service/outputs.tf:16]: Let's split scheduled example into `worker-scheduled`?

**Review by DimmKirr — CHANGES_REQUESTED** [review]
Looks good, but some notes.
1. I would move acm cert to a separate example
2. Let's remove DNS from worker
3. Let's update subnets'/vpc's cidrs
4. Reduce locals usage
If you'd like let's discuss this and other comments 1:1 too.

File: examples/complete-worker/variables.tf (hunk 14 above)
Comment (igorkotof) [reply examples/complete-worker/variables.tf]: it takes time to test carefully  - let's update in separate story

File: examples/complete-web/data.tf [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    name   = "name"
+    values = ["amzn2-ami-ecs-hvm-*-x86_64-ebs"]
+  }
+
+  filter {
+    name   = "owner-alias"
+    values = ["amazon"]
+  }
+}
+
```
Comment (DimmKirr) [inline examples/complete-web/data.tf]: This seems not to be required as we don't use ec2 s
```suggestion
```

File: examples/complete-web/variables.tf [hunk 23]
```diff
@@ -0,0 +1,28 @@
+locals {
+  env                  = var.env
```
Comment (DimmKirr) [inline examples/complete-web/variables.tf]: Please let's not use locals when there is no actual need in them (we're not constructing the variable). Unless you have a reason to keep them.
(here, everywhere in this file and other examples)

File: examples/complete-worker/data.tf [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+  filter {
+    name   = "name"
+    values = ["amzn2-ami-ecs-hvm-*-x86_64-ebs"]
+  }
+
+  filter {
+    name   = "owner-alias"
+    values = ["amazon"]
+  }
+}
```
Comment (DimmKirr) [inline examples/complete-worker/data.tf]:
```suggestion
```

File: examples/web-nginx-proxy/data.tf [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+  filter {
+    name   = "name"
+    values = ["amzn2-ami-ecs-hvm-*-x86_64-ebs"]
+  }
+
+  filter {
+    name   = "owner-alias"
+    values = ["amazon"]
+  }
+}
```
Comment (DimmKirr) [inline examples/web-nginx-proxy/data.tf:18]:
```suggestion
```

File: examples/worker-scheduled/main.tf [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+  memory           = 2048
+  cpu              = 1024
+
+  # Containers
+  ecs_cluster_arn      = module.ecs.cluster_arn
+  docker_registry      = local.docker_registry
+  image_id             = local.image_id
+  docker_image_tag     = local.docker_image_tag
+
+  docker_container_command           = ["rake", "notify:daily"]
```
Comment (DimmKirr) [inline examples/worker-scheduled/main.tf]:
```suggestion
  docker_container_command           = ["echo", "here-is-the-output"]
```

File: examples/worker-scheduled/main.tf [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+    ENV      = var.env
+    APP_NAME = "App"
+  }
+
+  iam_role_policy_statement = [
+    {
+      Effect   = "Allow",
+      Action   = "s3:*",
+      Resource = "*"
+  }]
```
Comment (DimmKirr) [inline examples/worker-scheduled/main.tf]: Irrelevant
```suggestion
```

File: examples/worker-scheduled/main.tf [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+  root_domain_name = var.root_domain_name
+  zone_id          = local.zone_id
+
+  # Environment variables
+  app_secrets = [
+  ]
+  environment = {
+    ENV      = var.env
+    APP_NAME = "App"
+  }
```
Comment (DimmKirr) [inline examples/worker-scheduled/main.tf]: Irrelevant
```suggestion
```

File: examples/complete-worker/main.tf [hunk 29]
```diff
@@ -16,14 +69,12 @@ module "web_complete" {
...
 
   # Containers
+  ecs_cluster_arn      = module.ecs.cluster_arn
   docker_registry      = local.docker_registry
   image_id             = local.image_id
   docker_image_tag     = local.docker_image_tag
-  iam_instance_profile = local.iam_instance_profile
-  key_name             = local.key_name
 
   docker_container_command           = ["rake", "notify:daily"]
```
Comment (DimmKirr) [inline examples/complete-worker/main.tf]:
```suggestion
  docker_container_command           = ["echo", "command-output"]
```

**Review by DimmKirr — CHANGES_REQUESTED** [review]
Ignore locals for now

File: examples/web-nginx-proxy/data.tf (hunk 25 above)
Comment (igorkotof) [reply examples/web-nginx-proxy/data.tf:18]: @AutomationD we use it as image id for ECS-APP

File: examples/complete-worker/data.tf (hunk 24 above)
Comment (igorkotof) [reply examples/complete-worker/data.tf]: @AutomationD we use it as image id for ECS-APP

File: examples/complete-web/data.tf (hunk 22 above)
Comment (igorkotof) [reply examples/complete-web/data.tf]: @AutomationD we use it as image id for ECS-APP

**Review by DimmKirr — APPROVED** [review]
Looks good, but let's discuss ec2 image prior merging.

**Pull request merged**